src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Activity.cs

```diff
@@ -100,7 +121,78 @@ public async Task UpdateInstanceBlobsAsync([ActivityTrigger] UpdateInstanceBlobA
+                });
+
+            logger.LogInformation("Completed updating instance blobs starting with [{Start}, {End}]. Total batchSize {BatchSize}.",
+                batch[0],
+                batch[^1],
+                batchSize);
+
+            processed += batchSize;
+        }
+
+        logger.LogInformation("Completed updating all instance blobs");
+
+        // await UpdateInstanceBlobsV2Async(arguments, logger);
+    }
+
+    /// <summary>
+    /// Asynchronously batches the instance watermarks and calls the update instance.
+    /// </summary>
+    /// <param name="arguments">BatchUpdateArguments</param>
+    /// <param name="logger">A diagnostic logger.</param>
+    /// <returns>
+    /// A task representing the <see cref="UpdateInstanceBlobsV2Async"/> operation.
+    /// </returns>
+    /// <exception cref="ArgumentNullException">
+    /// <paramref name="arguments"/> or <paramref name="logger"/> is <see langword="null"/>.
+    /// </exception>
+    [FunctionName(nameof(UpdateInstanceBlobsV2Async))]
+    public async Task<IReadOnlyList<InstanceMetadata>> UpdateInstanceBlobsV2Async([ActivityTrigger] UpdateInstanceBlobArguments arguments, ILogger logger)
+    {
+        EnsureArg.IsNotNull(arguments, nameof(arguments));
+        EnsureArg.IsNotNull(arguments.ChangeDataset, nameof(arguments.ChangeDataset));
+        EnsureArg.IsNotNull(arguments.InstanceMetadatas, nameof(arguments.InstanceMetadatas));
+        EnsureArg.IsNotNull(arguments.Partition, nameof(arguments.Partition));
+        EnsureArg.IsNotNull(logger, nameof(logger));
+
+        DicomDataset datasetToUpdate = GetDeserialzedDataset(arguments.ChangeDataset);
+
+        int processed = 0;
+
+        logger.LogInformation("Beginning to update all instance blobs, Total count {TotalCount}", arguments.InstanceMetadatas.Count);
+
+        List<InstanceMetadata> updatedInstances = new List<InstanceMetadata>();
+        while (processed < arguments.InstanceMetadatas.Count)
+        {
+            int batchSize = Math.Min(_options.BatchSize, arguments.InstanceMetadatas.Count - processed);
+            var batch = arguments.InstanceMetadatas.Skip(processed).Take(batchSize).ToList();
+
+            logger.LogInformation("Beginning to update instance blobs for range [{Start}, {End}]. Total batch size {BatchSize}.",
+                    batch[0],
+                    batch[^1],
+                    batchSize);
+
+            await Parallel.ForEachAsync(
+                batch,
+                new ParallelOptions
+                {
+                    CancellationToken = default,
+                    MaxDegreeOfParallelism = _options.MaxParallelThreads,
+                },
+                async (instance, token) =>
+                {
+                    FileProperties fileProperties = await _updateInstanceService.UpdateInstanceBlobAsync(instance.ToInstanceFileState(), datasetToUpdate, arguments.Partition, token);
+                    updatedInstances.Add(
```

Comment:
You need this collection to be thread-safe. I am unsure if `List<T>` has thread-safe inserts; I don't think so, but please double-check. Otherwise use one of the "Concurrent" ones